### DISCUSSION TITLE
Documentation for DigitalOcean Spaces config?

### DISCUSSION THREAD

**Discussion by andrewgioia** [pixelfed/pixelfed#5119] [Support]
I'm trying to get DO Spaces setup as an s3 storage provider but so far cannot figure out how to get it working. The [configuration page](https://docs.pixelfed.org/technical-documentation/env.html#filesystem) lists the config variables but doesn't have any explanation, and [`filesystems.php`](https://github.com/pixelfed/pixelfed/blob/f5260902293b35c830c865f5bc88162ae6e60f57/config/filesystems.php) has some additional context but not a lot.

Is Spaces fully supported yet? If so, what changes would I need to make here to my current .env:

```
PF_ENABLE_CLOUD=true
FILESYSTEM_DRIVER=spaces
DO_SPACES_KEY={key}
DO_SPACES_SECRET={secret}
DO_SPACES_ENDPOINT=nyc3.digitaloceanspaces.com
DO_SPACES_REGION=nyc3
DO_SPACES_BUCKET="{spaces-name}"
DO_SPACES_ROOT=
```

Thanks, happy to provide any debugging info as needed.

**Comment by gergelypolonkai** [comment]
The problem seems to be that `filesystem.php` generates an URL that depends on the DO Spaces bucket configured with CDN enabled.

However, if i set up everything based on this, it tries to connect to a hostname `my-bucket` instead of `region.digitaloceanspaces.com`.

**Comment by BorisKourt** [comment]
Has anything changed around this in the last year? I am having a similar issue it seems.

**Comment by BorisKourt** [comment]
I was able to get it to work (as far as the current testing shows) doing the following:

```sh
## S3 Configuration (Post-Installer)
PF_ENABLE_CLOUD=true
FILESYSTEM_DRIVER=s3 # "spaces" driver doesn't work with Digital Ocean CDN
FILESYSTEM_CLOUD=s3
AWS_ACCESS_KEY_ID=<snip>
AWS_SECRET_ACCESS_KEY=<snip>
AWS_DEFAULT_REGION=us-east-1 # Doesn't have to be the DO region. This is an AWS region.
AWS_BUCKET=<snip>
AWS_ENDPOINT=https://ams3.digitaloceanspaces.com # Not the same endpoint as what DO links to in the bucket. Note the missing bucket name.
AWS_URL=https://your.cdn.subdomain.com # Since I am using the CDN this was the only way to do it. https:// is mandatory.
```

**Comment by mitexleo** [comment]
> Has anything changed around this in the last year? I am having a similar issue it seems.

Kindly share your diagnostics with us

**Comment by mitexleo** [comment]
Oh ... Sorry ..

**Comment by BorisKourt** [comment]
The spaces driver doesn't work with the CDN because there is no way to provide an equiv of the `AWS_URL` to the DO parameters. [filesystems](https://github.com/pixelfed/pixelfed/blob/dev/config/filesystems.php#L72-L85)

**Comment by BorisKourt** [comment]
This works for single image posts (haven't tried galleries). But for profile pictures the image path is malformed:

```
https://example.comhttps//cdn.example.com/public/avatars/051/405/995/096/096/768/1/kYSiafmPTyUrB4Jsjjlq_avatar.png?v=2
```

It is adding `https://example.com` in front of the `AWS_URL` value of `https//cdn.example.com` 

It should just be:

```
https//cdn.example.com/public/avatars/051/405/995/096/096/768/1/kYSiafmPTyUrB4Jsjjlq_avatar.png?v=2
```

So far I have only seen this for profile pictures, and not with any other uploads. 

And I can see the file at the correct path in the bucket.

Where in the codebase is this URL formed?
